sdk/metric/sdk.go

```diff
@@ -139,6 +142,10 @@ type (
 		labels        *label.Set
 		observed      export.Aggregator
 	}
+
+	// MetricsLabelsEnricher can be provided as a config option to enrich metrics labels based on
+	// the context when the metrics are recorded
+	MetricsLabelsEnricher func(context.Context, []label.KeyValue) ([]label.KeyValue, error)
```

Comment:
I would prefer this be

```
type LabelEnricher interface {
  Enrich(context.Context, label.Set) label.Set
}
```

I made the input be `label.Set` in order to support the bound instrument case, where the attributes have already been formed into a set, e.g., to support enriching these events:
```
func (r *record) RecordOne(ctx context.Context, num number.Number)
```

Reply:
sorry @jmacd I'm not sure I get this correctly, do you mean the enrichment should happen in `RecordOne` method on the `record` instead of the one one `syncInstrument`?

Reply:
@hstan I am sorry about the delay. You pointed out that `label.Set` can't be copied, and I had been thinking about the bound-instrument case where we can't modify the `label.Set`. I had to think about this some more.

For unbound instruments, `acquireHandle()` could enrich the labels "at once". Currently there is a call to `label.NewSetWithSortable` that sorts and de-duplicates call-site labels and returns a new `label.Set`, and we can arrange to call the enricher on the context and combine result, meaning only a single sort-and-deduplicate pass.

The bound instrument code path has `RecordOne()` called with a context after `acquireHandle()` was already called. This means the `*record` being referred to likely won't be used when an enricher is in use (a number of optimizations will suggest themselves here, later), but we have its `label.Set`. The problem is that `label.Set` doesn't give you a `[]label.KeyValue` we could use to call the same `label.Set` constructor (due to support for `label.Distinct`). I'm having trouble finding an interface that doesn't require a lot of copying when the enricher is used with bound instruments, but I don't think this is a real concern. To ensure bound instruments are handled correctly, the `(*record).RecordOne()` method will have to change a bit when there's an enricher. I imagine a "reacquireHandle()" method there's an enricher: in this code path use the bound instrument's `label.Set.ToSlice()`, then call the enricher with the context, then follow the unbound instrument code path if any enrichment occurs.

Later if performance is a sore point, it may require changes in the enrichment API or it may not, we'll see. Thanks for this and again sorry for the long delay!

Reply:
This is a pretty confusing comment. I'm going to resolve it and re-state it below.